hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -374,15 +375,18 @@ def fill_for(elements, unique, fill, name=""):
     return fill
 
 
+D = TypeVar("D", bound=DTypeLike)
+
+
 @defines_strategy(force_reusable_values=True)
 def arrays(
-    dtype: Any,
+    dtype: D,
     shape: Union[int, st.SearchStrategy[int], Shape, st.SearchStrategy[Shape]],
     *,
     elements: Optional[Union[st.SearchStrategy, Mapping[str, Any]]] = None,
     fill: Optional[st.SearchStrategy[Any]] = None,
     unique: bool = False,
-) -> st.SearchStrategy[np.ndarray]:
+) -> st.SearchStrategy[NDArray[D]]:
```

Comment:
We could also declare the shape here, for constant shapes... albeit by going from two to four overloads (value or strategy for each of two arguments).

Reply:
I think we can get away without overloads in this case, by defining another typevar for shape but honestly I would like to leave this point for now until we sort out the rest the comments you made in this PR.

Reply:
Sounds good to me, I'll be very happy to get in whatever improvement we have and write up an issue with the rest.  👍